contributors/design-proposals/storage/volume-topology-scheduling.md

```diff
@@ -78,16 +78,158 @@ pod placement, volume binding and provisioning should be more tightly coupled wi
+}
+
+type VolumeNodeAffinity struct {
+    // The PersistentVolume can only be accessed by Nodes that meet
+    // these required constraints
+    Required *NodeSelector
+}
+```
+
+The `Required` field is a hard constraint and indicates that the PersistentVolume
+can only be accessed from Nodes that satisfy the NodeSelector.
+
+In the future, a `Preferred` field can be added to handle soft node contraints with
+weights, but will not be included in the initial implementation.
+
+The advantages of this NodeAffinity field vs the existing method of using zone labels
+on the PV are:
+* We don't need to expose first-class labels for every topology domain.
+* Implementation does not need to be updated every time a new topology domain
+  is added to the cluster.
+* NodeSelector is able to express more complex topology with ANDs and ORs.
+
+Some downsides include:
+* You can have a proliferation of Node labels if you are running many different
+  kinds of volume plugins, each with their own topology labeling scheme.
+
+
+### Example PVs with NodeAffinity
+#### Local Volume
+```
+apiVersion: v1
+kind: PersistentVolume
+metadata:
+  Name: local-volume-1
+spec:
+  capacity:
+    storage: 100Gi
+  storageClassName: my-class
+  local:
+    path: /mnt/disks/ssd1
+  nodeAffinity:
+    required:
+      nodeSelectorTerms:
+      - matchExpressions:
+        - key: kubernetes.io/hostname
+          operator: In
+          values:
+          - node-1
+```
+
+#### Zonal Volume
+```
+apiVersion: v1
+kind: PersistentVolume
+metadata:
+  Name: zonal-volume-1
+spec:
+  capacity:
+    storage: 100Gi
+  storageClassName: my-class
+  gcePersistentDisk:
+    diskName: my-disk
+    fsType: ext4
+  nodeAffinity:
+    required:
+      nodeSelectorTerms:
+      - matchExpressions:
+        - key: failure-domain.beta.kubernetes.io/zone
+          operator: In
+          values:
+          - us-central1-a
+```
+
+#### Multi-Zonal Volume
+```
+apiVersion: v1
+kind: PersistentVolume
+metadata:
+  Name: multi-zonal-volume-1
+spec:
+  capacity:
+    storage: 100Gi
+  storageClassName: my-class
+  gcePersistentDisk:
+    diskName: my-disk
+    fsType: ext4
+  nodeAffinity:
+    required:
+      nodeSelectorTerms:
+      - matchExpressions:
+        - key: failure-domain.beta.kubernetes.io/zone
+          operator: In
+          values:
+          - us-central1-a
+          - us-central1-b
+```
 
-## Design
+### Default Specification
+Existing admission controllers and dynamic provisioners for zonal volumes
+will be updated to specify PV NodeAffinity in addition to the existing zone
+and region labels.  This will handle newly created PV objects.
+
+Existing PV objects will have to be upgraded to use the new NodeAffinity field.
+This does not have to occur instantaneously, and can be updated within the
+deprecation period.
+
+TODO: This can be done through one of the following methods:
+- Manual updates/scripts
+- cluster/update-storage-objects.sh?
+- A new PV update controller
```

Comment:
This should be automatic unless absolutely impossible?

Reply:
Are you saying that there is already a general object upgrade/backfill mechanism that exists in Kubernetes today?  As far as I know, there isn't, so if we want it to be automatic, then the best way I can think of is to add a PV update controller.

Reply:
Just had a discussion with @caesarxuchao.  api-machinery is working on an update controller, however it's for updating objects between two different versions.  Since this is just adding a new field in the same version, then their solution won't work for us.

So we discussed a possible solution for our use case:
- PV update controller to copy the zone/region labels from existing PVs and specify them in the new field
- Admission controller to reject conflicting updates (ie changing the zone in the label, but not the field)